### PR TITLE
[tflite] add same scale constraint to tflite's mean op

### PR DESCRIPTION
**Pull request by freedomtan**
Add SameOperandsAndResultsScal to tflite's mean. Without this constraint, post-training quantization ([PTQ](https://www.tensorflow.org/lite/performance/post_training_quantization)) may result in MEAN with different quantization scales for inputs and outputs. Because NNAPI only supports MEAN with same quantization parameters for inputs and output, this kind of mean ops could not be delegated to NNAPI via the NNAPI delegate.

E.g., if we do PTQ on the [ResNet 50 from tfhub](https://tfhub.dev/tensorflow/resnet_50/classification/1),
```python
import itertools
import os
import pathlib
import requests
import tarfile

import tensorflow as tf
import tensorflow_datasets as tfds

saved_model_dir = "./resnet_saved_model/"
resnet_saved_model_file = "https://tfhub.dev/tensorflow/resnet_50/classification/1?tf-hub-format=compressed"
response = requests.get(resnet_saved_model_file, stream=True)
file = tarfile.open(fileobj=response.raw, mode="r|gz")
file.extractall(path=saved_model_dir)

imagenet_validation = tfds.load(name="imagenet2012", split="validation")

def representative_data_gen():
  for imagenet_example in imagenet_validation.take(100):
    image, label = imagenet_example["image"], imagenet_example["label"]
    image = tf.cast(image, tf.float32) / 255.0
    image = tf.image.central_crop(image, central_fraction=0.875)
    image = tf.expand_dims(image, 0)
    image = tf.image.resize(image, (224,224))
    yield [image]

converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_data_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.int8 
converter.inference_output_type = tf.int8 
tflite_quant_model = converter.convert()

tflite_quant_model_file = pathlib.Path('/tmp')/"resnet50_quant.tflite"
tflite_quant_model_file.write_bytes(tflite_quant_model)
```
then `adb push /tmp/resnet50_quant.tflite /data/local/tmp/`, and run `benchmark_model` on device (here I ran it on a Pixel 4)
```
$ ./benchmark_model_validation --graph=resnet50_qnant.tflite   --use_nnapi=1 --enable_op_profiling=1

```
It shows something like:
```
STARTING!
Log parameter values verbosely: [0]
Graph: [resnet_v15_quant_original_mean.tflite]
Enable op profiling: [1]
Use NNAPI: [1]
NNAPI accelerators available: [qti-default,qti-dsp,qti-gpu,google-edgetpu,nnapi-reference]
Loaded model resnet50_quant.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite delegate for NNAPI.
NNAPI delegate created.
WARNING: Operator MEAN (v2) refused by NNAPI delegate: NNAPI requires that the input and output have the same quantization parameters.
INFO: Replacing 75 node(s) with delegate (TfLiteNnapiDelegate) node, yielding 3 partitions.
Explicitly applied NNAPI delegate, and the model graph will be partially executed by the delegate w/ 2 delegate kernels.
.......
Operator-wise Profiling Info for Regular Benchmark Runs:
============================== Run Order ==============================
	             [node type]	          [start]	  [first]	 [avg ms]	     [%]	  [cdf%]	  [mem KB]	[times called]	[Name]
	     TfLiteNnapiDelegate	            0.009	   14.431	   14.449	 48.545%	 48.545%	     0.000	        1	[resnet50/activation_48/Relu;resnet50/add_15/add]:76
	                    MEAN	           14.460	   13.399	   13.501	 45.359%	 93.904%	     0.000	        1	[resnet50/reduce_mean/Mean]:73
	     TfLiteNnapiDelegate	           27.962	    1.754	    1.814	  6.096%	100.000%	     0.000	        1	[StatefulPartitionedCall:0]:77
.......
Number of nodes executed: 3
============================== Summary by node type ==============================
	             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
	     TfLiteNnapiDelegate	        2	    16.263	    54.642%	    54.642%	     0.000	        2
	                    MEAN	        1	    13.500	    45.358%	   100.000%	     0.000	        1

Timings (microseconds): count=50 first=29584 curr=22585 min=22585 max=33233 avg=29764 std=1779
Memory (bytes): count=0
3 nodes observe
```
With this constraint applied, for the same ResNet 50 model, I got something like the following:
```
....
Number of nodes executed: 1
============================== Summary by node type ==============================
	             [Node type]	  [count]	  [avg ms]	    [avg %]	    [cdf %]	  [mem KB]	[times called]
	     TfLiteNnapiDelegate	        1	    14.762	   100.000%	   100.000%	     0.000	        1

Timings (microseconds): count=67 first=14774 curr=15087 min=13959 max=15226 avg=14762.4 std=248
Memory (bytes): count=0
1 nodes observed
```

### CODE REVIEW

**Comment by freedomtan** [comment]
@multiverse-tf and @miaowang14 For your information, I ran into some PTQ models with TFLite MEAN v2 which is not supported NNAPI.

**Review by abattery — APPROVED** [review]

**Comment by teijeong** [comment]
Since TFLite kernel supports mean op with different input and output  scales, this would solve NNAPI issue but might introduce problems for other models' performance. I think this fix will surely be an easiest way to get models running with NNAPI, but more complex but safer fix would be needed to prevent regressions.

**Comment by abattery** [comment]
@miaowang14  Instead of failing on the NNAPI side, can we reject the partition when there are unsupported cases in the Mean op in the partitioning logic in the NNAPI delegate?

**Comment by freedomtan** [comment]
> @miaowang14  Instead of failing on the NNAPI side, can we reject the partition when there are unsupported cases in the Mean op in the partitioning logic in the NNAPI delegate? 

@abattery The MEAN op is rejected by the NNAPI delegate now (not by the NNAPI runtime or drivers)

**Comment by abattery** [comment]
I see. Thanks @freedomtan for the quick confirmation.

There will be something we can do when we would like to convert the given graph based on the related circumstances. @renjie-liu Please take a look.

**Comment by freedomtan** [comment]
@teijeong, @abattery and @renjie-liu As far as I can tell, yes, TFLite MEAN kernel supports mean op with different input and output scales, but that means implicit re-quantization is done inside the kernel. My guess is that there is no performance lost when scales are the same. I guess if proper logic is added into kernel, same quantization scale should be a bit faster. There could be potential accuracy issues though.

| original | after this one-line patch|
| ---------- | -------------------------------|
| |  
|

**Comment by renjie-liu** [comment]
I think there are a few things:

- It's not guaranteed mean op should have same scale for both inputs & outputs, imagine you have an input tensor [-3.0, 0.0, 3.0], the reduced output will be just [0.0] (the scale shrunk down!) 
- So Ensure the input output have the same scale does not help on keeping accuracy (and it hurt the accuracy)
- Well, it does not simplify the implementation either: For reduce mean, you will still need to do at least one "division" : sum(elements) / element_count, and you actually fuse the input-output scale withe the element_count, see how we implemented [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/optimized/integer_ops/mean.h#L210-L213) 
- TfLite CPU kernel supports different scales for mean op.

To sum up, I don't think we should add the constraints to the Mean op for the reasons mentioned above, instead, we should fix on the NNAPI side.

If you need a temporary workaround, I think you should go for reduce_sum then a broadcast_mul with (reciprocal of tensor shape).

**Comment by freedomtan** [comment]
@renjie-liu Agree there could be accuracy problem, but I don't see why shrunk scale is better. We know MEAN (different from some common cases, such as CONCAT) is a special case that same scale constraint for inputs and outputs is guaranteed to work most of time.

**Comment by renjie-liu** [comment]
You can consider `mean` is more like a "smoother" operation, so `mean` can shrink the output range (a lot).

I agree that maybe let `mean` have the same constraint for inputs and outputs can work sometime, but I don't think that can work for all cases.

`avg_pool` may have this constraint since most of the time the window size is not large, `mean` is more like a global `avg_pool` so the situation may change.

On the safe side, I think we should keep the current setting for `mean`, we probably can push NNAPI to fix the issue.

As for the temporary workaround, I think you can either do: reduce_sum then divide the element_count or you can just transform the mean op into a global avg_pool.

**Comment by freedomtan** [comment]
Thanks for the comments. I'll close this and open an issue instead later.
